File: hilti/python/hilti/codegen/channel.py
# $Id$ 
#
# Code generation for the channel type.

import llvm.core

from hilti.core import *
from hilti import instructions
from codegen import codegen

_doc_c_conversion = """
A ``channel`` is mapped to a ``hlt_channel *``. The type is defined in
|hilti_intern.h|:

.. literalinclude:: /libhilti/hilti_intern.h
   :start-after: %doc-hlt_channel-start
   :end-before:  %doc-hlt_channel-end
   
"""

@codegen.typeInfo(type.Channel)
def _(type):
    typeinfo = codegen.TypeInfo(type)
    typeinfo.c_prototype = "hlt::channel *"
    typeinfo.to_string = "hlt::channel_to_string";
    return typeinfo

@codegen.llvmCtorExpr(type.Channel)
def _(op, refine_to):
    assert False

@codegen.llvmType(type.Channel)
def _(type, refine_to):
    return codegen.llvmTypeGenericPointer()
       
@codegen.operator(instructions.channel.New)
def _(self, i):
    top = instruction.TypeOperand(i.op1().value())
    result = self.llvmGenerateCCallByName("hlt::channel_new", [top], [top.type()])
    self.llvmStoreInTarget(i.target(), result)

@codegen.when(instructions.channel.Write)
def _(self, i):
    self.llvmGenerateCCallByName("hlt::channel_try_write", [i.op1(), i.op2()], [i.op1().type(), i.op2().type()])

@codegen.when(instructions.channel.Read)
def _(self, i):
    voidp = self.llvmGenerateCCallByName("hlt::channel_try_read", [i.op1()], [i.op1().type()])
    item_type = i.op1().type().refType().itemType()
    nodep = self.builder().bitcast(voidp, llvm.core.Type.pointer(self.llvmTypeConvert(item_type)))
    self.llvmStoreInTarget(i.target(), self.builder().load(nodep))

@codegen.when(instructions.channel.Size)
def _(self, i):
    result = self.llvmGenerateCCallByName("hlt::channel_get_size", [i.op1()], [i.op1().type()])
    self.llvmStoreInTarget(i.target(), result)

File: hilti/codegen/instructions/old/channel.cc

#include <hilti.h>

#include "../stmt-builder.h"

using namespace hilti;
using namespace codegen;


void StatementBuilder::visit(statement::instruction::channel::Read* i)
{
    /*
        auto op1 = cg()->llvmValue(i->op1(), X);

        auto result = builder()->

        cg()->llvmStore(i, result);
    */

    /*
        CodeGen::expr_list args;
        args.push_back(i->op1());

        auto result = cg()->llvmCall("hlt::X", args);

        cg()->llvmStore(i, result);
    */


    /*
        def _codegen(self, cg):
            self.blockingCodegen(cg)

    */
}

void StatementBuilder::visit(statement::instruction::channel::ReadTry* i)
{
    /*
        auto op1 = cg()->llvmValue(i->op1(), X);

        auto result = builder()->

        cg()->llvmStore(i, result);
    */

    /*
        CodeGen::expr_list args;
        args.push_back(i->op1());

        auto result = cg()->llvmCall("hlt::X", args);

        cg()->llvmStore(i, result);
    */


    /*
        def _codegen(self, cg):
            voidp = cg.llvmCallC("hlt::channel_read_try", [self.op1()])
            item_type = self.target().type()
            nodep = cg.builder().bitcast(voidp, llvm.core.Type.pointer(cg.llvmType(item_type)))
            cg.llvmStoreInTarget(self, cg.builder().load(nodep))

    */
}

void StatementBuilder::visit(statement::instruction::channel::Size* i)
{
    /*
        auto op1 = cg()->llvmValue(i->op1(), X);

        auto result = builder()->

        cg()->llvmStore(i, result);
    */

    /*
        CodeGen::expr_list args;
        args.push_back(i->op1());

        auto result = cg()->llvmCall("hlt::X", args);

        cg()->llvmStore(i, result);
    */


    /*
        def _codegen(self, cg):
            result = cg.llvmCallC("hlt::channel_size", [self.op1()])
            cg.llvmStoreInTarget(self, result)

    */
}

void StatementBuilder::visit(statement::instruction::channel::Write* i)
{
    /*
        auto op1 = cg()->llvmValue(i->op1(), X);
        auto op2 = cg()->llvmValue(i->op2(), X);
    */

    /*
        CodeGen::expr_list args;
        args.push_back(i->op1());
        args.push_back(i->op2());
        cg()->llvmCall("hlt::X", args);
    */


    /*
        def _codegen(self, cg):
            self.blockingCodegen(cg)

    */
}

void StatementBuilder::visit(statement::instruction::channel::WriteTry* i)
{
    /*
        auto op1 = cg()->llvmValue(i->op1(), X);
        auto op2 = cg()->llvmValue(i->op2(), X);
    */

    /*
        CodeGen::expr_list args;
        args.push_back(i->op1());
        args.push_back(i->op2());
        cg()->llvmCall("hlt::X", args);
    */


    /*
        def _codegen(self, cg):
            op2 = self.op2().coerceTo(cg, self.op1().type().refType().itemType())
            cg.llvmCallC("hlt::channel_write_try", [self.op1(), op2])

    */
}
